### PR TITLE
Replace Slice= dependencies when specified more than once

### LINKED ISSUE #20182: Non-deterministic Slice= assignment
**systemd version the issue has been seen with**
 > 249

**Used distribution**
 > openSUSE Tumbleweed (mkosi image)

**Linux kernel version used** (`uname -a`)
 > `5.13.0-rc7`

**Expected behaviour you didn't see**
 > The slice unit configured for a unit follows deterministic rules (e.g. last assignment wins as with <v249).

**Unexpected behaviour you saw**
 > Slice assignments are evaluate in random order

```
image:~ # cat /etc/systemd/system/foo.service
[Service]
Slice=bar.slice
Slice=bar2.slice
ExecStart=/usr/bin/sleep inf

image:~ # cat /etc/systemd/system/goo@.service
[Service]
Slice=car.slice
Slice=car2.slice
ExecStart=/usr/bin/sleep inf

Control group /:
-.slice
├─bar.slice 
│ └─foo.service 
│   └─174 /usr/bin/sleep inf
├─car2.slice 
│ └─goo@alpha.service 
│   └─198 /usr/bin/sleep inf
[...]
```

**Steps to reproduce the problem**
> - create a unit file (foo.service) with multiple `Slice=` assignments,
> - `systemctl daemon-reload`
> - `systemctl start foo.service`
> - watch the service lands where in the tree 

I noticed this when looking at d219a2b07cc5dc8ffd5010f08561fab2780d8616. At first, I suspected that the custom slice assignment may not be honored since it would share drawing drum with the default assignment. My quick test didn't show that but I also can't see why it couldn't happen with other unit/slice names (IIUC, hashmap ordering is pseudorandom). The confirmed issue is that it changes semantics of multiple `Slice=` assignments (the last one used to overwrite previous), custom slices _likely_ work. Nevertheless, the new deps implementation should allow replacing for one-of-a-kind relations (such as slice).

### PR DESCRIPTION
**Pull request by anitazha**
Fixes #20182

### CODE REVIEW

**Comment by poettering** [comment]
Hmm, I envisioned a simpler fix: instead of immediately synthesizing deps from the slice untis that are configured do so only once the files are fully loaded, i.e. in unit_load() where we add all the other automatic stuff too. That should be a much simpler approach as it doesn't require us to remove any deps we already created.

**Comment by anitazha** [comment]
> instead of immediately synthesizing deps from the slice untis that are configured do so only once the files are fully loaded, i.e. in unit_load()

Sure I can update this to be simpler for slices.

What's your stance on removing dependencies in general though? For something like #18154 it doesn't seem like it's as simple to delay synthesizing other dependencies like After=/Before= since they can have multiple units.

File: src/core/unit.c [hunk 1]
```diff
@@ -3317,6 +3393,10 @@ int unit_set_slice(Unit *u, Unit *slice, UnitDependencyMask mask) {
         if (UNIT_GET_SLICE(u) && u->cgroup_realized)
                 return -EBUSY;
 
+        /* Only allow setting one slice; remove previous ones if they exist */
+        if (UNIT_GET_SLICE(u) && UNIT_GET_SLICE(u) != slice)
+                unit_remove_dependencies_by_type(u, UNIT_IN_SLICE);
+
```
Comment (Werkov) [inline src/core/unit.c:3323]: `UNIT_GET_SLICE(u) != slice` is unnecessary here (given the bail out on `UNIT_GET_SLICE(u) == slice` above.

**Comment by keszybz** [comment]
I'll set the reviewed/… label, since @anitazha said she'll post another version.

**Comment by anitazha** [comment]
Here's a version that delays adding slice dependencies until `unit_load()`. The change is a lot simpler but I'm not actually sure I got all the cases where we could be trying to get a unit's slice before it gets added to the dependencies in `unit_load()`.

**Comment by Werkov** [comment]
I see that the `unit_set_slice()` in `bus_unit_set_transient_property()` is not covered (the update is not followed by `unit_load()`).
IIRC, the DBus API would allow setting/updating the `Slice=` multiple times before the unit is actually started, so the trick with `unit_load()` once won't work (easily).

**Comment by poettering** [comment]
Hmm, I am very sorry. I had a closer look now and I actually think your previous approach was the way to go. Sorry for the confusion. I forgot how this worked. But it makes no sense to add a secondary field to the unit that configures which slice to use if we already have one in the UNIT_IN_SLICE dependency type, and we can nowadays delete deps already.

So here's what I propose: define a new bit in `UnitDependencyMask` that is set for deps that are the immediate result of a configured slice, i.e. are set for everything configured in Slice=. (`UnitDependencyMask` bitmask indicates how a dep came to be, i.e. what caused it to be created. we use it to be able to delete deps again by grouping all deps we might want to delete under one of these bits).

Once that new bit mask is defined use it for the UNIT_IN_SLICE dependency that is created in unit_set_slice(). With that in place we can remove all UNIT_IN_SLICE deps with a simple call to unit_remove_dependencies() with the flag passed.

I suspect the resulting patch should then be pretty short: it should suffice to adjust the mask when creating the deps, and then remove via that when changing the slice.

Sorry again for the confusion, I shouldn#t have wasted your time with the earlier review. sorry!

**Comment by anitazha** [comment]
No problem; third time's the charm? I took the suggestion to make an new UnitDependencyMask type and used that to remove older slices. Added a small test too.

**Comment by keszybz** [comment]
Looks nice.

**Pull request merged**